README.md
Introduction
============

This is a proof-of-concept implementation developed to explore issues relevant
to the redesign of Globo.com platform infra-structure.


Requirements for the New Platform
=================================

  Backend as a Service (BaaS) composed by cloud-based RESTful services: 
    * unified CMA
    * uniform access to all enterprise data (CMS,Search, Semantics)
    * authentication
    * authorization
    * billing

  The new architeture should naturally lead to support for mobile apps

  API Requests (80%) should respond below 20ms.


Objectives for this project
===========================

 1) Design a draft of the new architecture for content publication.
    Define a first version of the interfaces and contracts between the
    several components of the architecture.
    There is no need for a complete implementation at the end of the project.
    Development of code will be done on-demand to validate concepts.

 2) The design of the architecture will define the communication between the teams.
    We should be able to understand which stories are relevant to which teams. 

  The outcome of this project should be to deliver a draft implementation that
  does CMA+CDA for some content using the base elements of new architecture.
  The focus will be to elucidate doubts about the new architecture.

Data Model
==========

  Review <----> Software <-----> Categoria
     |                 \----------> Rating
     |
     \- Title
      \- Comments -> Comment

References
==========

 * https://github.com/globocom/mysqlapi

Desired features
================

  * the definiton of data structures shoul carry its own documentation
  * versioning (which level: API, resource, ...)
  * Role-based access control (RBAC) model for authorization profiles 

Configuration
=============

We applied two configurations in ElasticSearch:
 - elasticsearch/config/default-mapping.json 
 - elasticsearch/config/elasticsearch.yml

This files are inside the mythical-poc project tree, below config/elasticsearch. 
The goal of the configuration is to define that "uid" field always has exact match.

Execution
=========

cd ./jsonform/

python -m SimpleHTTPServer

Add Data
--------

Adding data to a schema-instance database
--

In our project, we distinguish adding instances (data) or schema (metadata).

Given the difference between classes and individuals in database models like relational
databases or a triplestore based on OWL, or, in other terms, given that classes are
first-class citizens in these models, we have two distinct data insertion points to desambiguate
these two aspects.

Adding schemas
---



Adding instances
---

jisadssjidsa


Retrieving data
------------

LALLALALALAL

```bash
curl -i -XPUT  'http://localhost:5100/data/schemas/software' -H 'Content-type: application/json' -T software_schema.json
```

```bash
curl -i -XPUT  'http://localhost:5100/data/schemas/review' -H 'Content-type: application/json' -T review_schema.json
```

```bash
curl -i -XPOST  'http://localhost:5100/data/softwares' -H 'Content-type: application/json' -T software_instance.json
```

Query textual data
------------------
curl -i -X GET "http://localhost:5100/data/reviews?q=\"windows\""

List supported query languages
------------------------------

curl -i -XGET 'http://localhost:5100/data/query'


Query using Cypher
------------------
  curl -i -X GET  "http://localhost:5100/data/query/cypher" -H "Content-type: application/json" -T cypher_query.json

Query using Gremlin
-------------------
  curl -i -X GET  "http://localhost:5100/data/query/gremlin" -H "Content-type: application/json" -d "g.v(37).map()"

Clean-up ElasticSearch
----------------------
  curl -X DELETE http://localhost:9200/data

Query ES
--------
  curl -XGET 'http://localhost:9200/data/_search?q=uid:"/data/schemas/software"&pretty=true'

Clean-up Neo4j
--------------
  start no=relationship(*) delete no;
  start no=node(*) delete no;

Setup before using Virtuoso
---------------------------
  create graph <http://mythical_poc.globo.com>


Deprecated
==========

 curl -i -XPOST 'http://localhost:5100/data/reviews' -H "Content-Type: application/json"  -d '{"title":"Novissimo Bla"}'
 curl -i -XPOST 'http://localhost:5100/schemas' -H "Content-Type: application/json"  -T "schemas/review_schema.json"


RELEASES.md
Releases
========

1. ✓ Fluxo completo para titulo de review (API/Python + CMA/Javascript + CDA/Rails + ElasticSearch) + CORS
2. ✓ Busca de uma review por título (caso de altíssima freqüência de leitura)
3. ✓ Suporte a dois bancos de dados (grafo/Neo4j + documentos/ES) mas schema+form continuam integrados 
4. ✓ Suporte ao relacionamento describedby que vincula instância a seu schema
5. ✓ Insercão de instâncias de software e review (relacionadas)
6. ✓ Introdução de consultas ao banco de grafos
7. ✓ Exemplo completo do TechTudo integrado com consultas


data/load_multiple.py
# -*- coding: utf-8 -*-
import json
import requests
import time


headers = {'Content-type': 'application/json'}


def load_json(filename):
    fp = open(filename, "r")
    return json.load(fp)


def load_string(filename):
    fp = open(filename, "r")
    return fp.read()


def wipeout_armageddon():
    print("Clean up databases")
    requests.post('http://localhost:7474/db/data/cypher', data='{"query": "start no=relationship(*) delete no"}', headers=headers)
    requests.post('http://localhost:7474/db/data/cypher', data='{"query": "start no=node(*) delete no"}', headers=headers)
    requests.delete('http://localhost:9200/data')

def load_schemas():
    print("Load schemas")
    requests.put('http://localhost:5100/data/schemas/software', data=load_string('software_schema.json'), headers=headers)
    requests.put('http://localhost:5100/data/schemas/software_category', data=load_string('software_category_schema.json'), headers=headers)
    requests.put('http://localhost:5100/data/schemas/comment', data=load_string('comment_schema.json'), headers=headers)
    requests.put('http://localhost:5100/data/schemas/review', data=load_string('review_schema.json'), headers=headers)


def load_instances():

    print("Load instances")
    instances = []

    ## Software Categories
    sample_category = {'$schema':'/data/schemas/software_category', 'name':'', 'items':[]}
    for category_name, slug in [(u'S.O.', 'SO'),
                          (u'Utilitátio', 'Utility'),
                          (u'Produtividade', 'Productivity'),
                          (u'Audio/Video', 'AudioVideo'),
                          (u'Internet', 'Internet'),
                          (u'Segurança', 'Security')]:
        sample_category['name'] = category_name
        response = requests.put('http://localhost:5100/data/software_categories/{0}'.format(slug),
                                data=json.dumps(sample_category), headers=headers)
        software_category_uid = '/data/software_categories/{0}'.format(slug)
        print("Software category instance " + software_category_uid)
        instances.append(software_category_uid)
    
    ## Softwares
    software_dict = load_json('software_instance.json')

    response = requests.post('http://localhost:5100/data/softwares', data=json.dumps(software_dict), headers=headers)
    software_uid = response.headers['Location']
    print("Software instance " + software_uid)
    instances.append(software_uid)

    software_dict['name'] = 'Linux'    
    response = requests.post('http://localhost:5100/data/softwares', data=json.dumps(software_dict), headers=headers)
    software_uid_linux = response.headers['Location']
    print("Software instance " + software_uid_linux)
    instances.append(software_uid_linux)

    software_dict['name'] = 'WSO2'
    software_dict['category'] = '/data/software_categories/Internet'
    response = requests.post('http://localhost:5100/data/softwares', data=json.dumps(software_dict), headers=headers)
    software_uid_wso2 = response.headers['Location']
    print("Software instance " + software_uid_wso2)
    instances.append(software_uid_wso2)

    ## Review Win8
    review_json = load_json('review_win8_instance.json')
    review_json['software'] = software_uid
    response = requests.post('http://localhost:5100/data/reviews', data=json.dumps(review_json), headers=headers)
    review_uid_win8 = response.headers['Location']
    print("Review instance " + review_uid_win8)
    instances.append(review_uid_win8)

    ## Review Linux
    review_json['software'] = software_uid_linux
    review_json['title'] = 'About Linux'
    response = requests.post('http://localhost:5100/data/reviews', data=json.dumps(review_json), headers=headers)
    review_uid_linux = response.headers['Location']
    print("Review instance " + review_uid_linux)
    instances.append(review_uid_linux)

    ## Review WSO2
    review_json['software'] = software_uid_wso2
    review_json['title'] = 'WSO2: yes or no?'
    response = requests.post('http://localhost:5100/data/reviews', data=json.dumps(review_json), headers=headers)
    review_uid_wso2 = response.headers['Location']
    print("Review instance " + review_uid_wso2)
    instances.append(review_uid_wso2)

    ## Comments
    comment1_json = load_json('comment1_win8_instance.json')
    comment1_json['context'] = review_uid_win8
    response = requests.post('http://localhost:5100/data/comments', data=json.dumps(comment1_json), headers=headers)
    comment1_uid = response.headers['Location']
    print("Comment1 instance " + comment1_uid)
    instances.append(comment1_uid)

    comment2_json = load_json('comment2_win8_instance.json')
    comment2_json['context'] = review_uid_win8
    comment2_json['reply'] = comment1_uid
    comment2_json['body'] = 'Oi tudo bem?'
    response = requests.post('http://localhost:5100/data/comments', data=json.dumps(comment2_json), headers=headers)
    comment2_uid = response.headers['Location']
    print("Comment2 instance " + comment2_uid)
    instances.append(comment2_uid)

    comment2_json = load_json('comment2_win8_instance.json')
    comment2_json['context'] = review_uid_win8
    comment2_json['reply'] = comment2_uid
    comment2_json['body'] = 'Belex'
    response = requests.post('http://localhost:5100/data/comments', data=json.dumps(comment2_json), headers=headers)
    comment2_uid = response.headers['Location']
    print("Comment2 instance " + comment2_uid)
    instances.append(comment2_uid)

    comment2_json['context'] = review_uid_linux
    comment2_json['body'] = 'Linux rulz!'
    del comment2_json['reply'] 
    response = requests.post('http://localhost:5100/data/comments', data=json.dumps(comment2_json), headers=headers)
    comment2_uid = response.headers['Location']
    print("Comment3 instance " + comment2_uid)
    instances.append(comment2_uid)

    return instances


def retrieve_data(path_name):
    response = requests.get('http://localhost:5100/data/{0}'.format(path_name), headers=headers)
    print("Retrieve {0}".format(path_name))
    print(response.json())

if __name__ == '__main__':
    wipeout_armageddon()
    load_schemas()
    paths = load_instances()
    retrieve_data('schemas/software')
    retrieve_data('schemas/review')
    retrieve_data('schemas/comment')
